### DISCUSSION TITLE
Questions about isAttributeEditAllowed function.

### DISCUSSION THREAD

**Discussion by NickWu** [Autodesk/maya-usd#2212] [Q&A] [answered]
Hi, I found some of the logic in `isAttributeEditAllowed` function a bit confusing https://github.com/Autodesk/maya-usd/blob/166a2ba7ef9216808ad35e8ae23362e1f7225746/lib/mayaUsd/ufe/Utils.cpp#L378

The following line of code checks the edit target layer’s position in the given prim: 
https://github.com/Autodesk/maya-usd/blob/166a2ba7ef9216808ad35e8ae23362e1f7225746/lib/mayaUsd/ufe/Utils.cpp#L390

But if it couldn’t find the layer in the prim’s layer stack, it still returns a position number and still check the prim’s property’s layer stack 
https://github.com/Autodesk/maya-usd/blob/166a2ba7ef9216808ad35e8ae23362e1f7225746/lib/mayaUsd/ufe/Utils.cpp#L410

But if a layer is not in a prim’s layer stack, it certainly will not in the prim’s property’s layer stack.  So I’d think when the edit target layer is not in a prim’s layer stack, `isAttributeEditAllowed` function could return true without further property stack checking so that a future prim spec for that prim can be contributed to the edit target layer. 

We found this issue just recently in maya 2022 because in maya 2020 `isAttributeEditAllowed` is not used so our workflow in maya 2020 had been working fine until we begin to use maya 2022. In our workflow, we tend to set different edit target layer in different department when a stage is opened, for instance, in layout we would set it to a layout assembly layer  and in Modelling we'd set it to a global assembly layer. And layout assembly layers is stronger than global assembly layers. Then, we'd move prims around and as a result write xforms into edit target layers. It might be adding a new prim spec or editing an existing prim spec. With` isAttributeEditAllowed`'s current logic, we won't be able to add a new prim spec if a prim doesn't have a prim spec in the edit target layer already. 

From my understanding, the intention for `isAttributeEditAllowed` is to warn/stop users when users make changes to a layer that would be overwritten by a more strong opinion in another layer. I'd think if a prim doesn't have a prim spec in the edit target layer, we should allow it to be able to write a new prim spec. And the second time a prim spec is writing to the layer, and if there is indeed another stronger prim spec in another layer, we can warn/stop users with `isAttributeEditAllowed`. What do you think?

**Comment by neilh-adsk** [comment, marked as answer]
Hi Nick,

I looked a little more into this and I tend to agree with the things you say.  Its always difficult when it comes to the implied restrictions imposed and I don't have a strong "opinion" on what to enforce with certain workflows.  It clearly is not a one size fits all approach.  Saying that in this case I think the `findLayerIndex` code non intuitive in that it returns a position which then causes the additional search in `isAttributeEditAllowed` which can cause the `false` condition.  So in this case I think it can be relaxed but would need to do so more verification to see if it breaks any other logic.  As a group we are revisiting the strength of Options being imposed and I look forward to starting up this conversation again !

**Comment by NickWu** [comment]
Thank you Neil. This issue is currently a blocker in our maya 2022 pipeline at Animal Logic, so we are making an internal code change for this. I can send a PR that is similar to what we did internally and start the conversation from there.

**Comment by NickWu** [comment]
I have created a PR for this issue https://github.com/Autodesk/maya-usd/pull/2390.
